### ISSUE TITLE
SIMD matrix inversion, large performance delta between JIT and AOT

### ISSUE THREAD

**Issue by jonahwilliams** [dart-lang/sdk#41805]
I'm currently investigating whether Flutter could make use of Dart's SIMD support to improve the performance of common matrix operations in the framework. For matrix inversion, the JIT is able to run about 5-6 times faster using SIMD operations.  But in AOT mode the SIMD code runs in about the same time as the non-SIMD, and often slower. This slowdown is consistent across dart2native on a desktop and Flutter release mode on a Pixel 4.

The benchmark I'm using is https://github.com/jonahwilliams/flutter/commit/f089fcd9261bf548cffc11a2c8de804cce0987a7 . The SIMD matrix inversion code is based off of https://github.com/tc39/ecmascript_simd , but I have not validated it extensively.

The non-SIMD matrix inversion is taken from `vector_math`: https://github.com/google/vector_math.dart/blob/master/lib/src/vector_math/matrix4.dart#L1224

JIT (Pixel 4)
vector_math: 0.07 - 0.10 us per invert
SIMD: 0.01 - 0.02 us per invert

AOT (Pixel 4)
vector_math:  ~0.1 us per invert
SIMD:  ~0.1  us per invert

Of course, the JIT is going to be better at all sorts of optimizations, but it seems like the vector_math implementation is not as negatively impacted.  Of course, I could also be measuring it wrong too...

> Dart version 2.9.0 (build 2.9.0-5.0.dev 9c94f08410)

**Comment by mkustermann** [comment]
>  Of course, I could also be measuring it wrong too...

Inlining can make a world of a difference with this benchmark (JIT inlines `invert` and AOT does not). To make the benchmark more fair it can be changed to actually use the result of previous inversion in the following inversion:
```dart
  SimdMatrix4 result = SimdMatrix4.identity();
  for (var i = 0; i < kLimit; i++) {
    result = result.invert();
  }   
```
instead of
```dart
  var simd4Identity = SimdMatrix4.identity();
  SimdMatrix4 result = SimdMatrix4.identity();
  for (var i = 0; i < kLimit; i++) {
    result = simd4Identity.invert()
  }   
```
That should explain the JIT AOT difference.

The AOT compiles the big `invert` function into around 200 instructions on arm64. So naively assuming 1 nanosecond per instruction would mean around 200 ns (i.e. 0.2 us) per inversion.

We could look into seeing if we can make the code a little better on arm64. Though I wouldn't expect a world of a difference. This 4x4 matrix inversion algorithm uses just a lot of operations.

**Comment by mraleph** [comment]
In this loop:

```
    var simd4Identity = SimdMatrix4.identity();
    for (var i = 0; i < 10000; i++) {
      result = simd4Identity.invert();
    }
```

If `simd4Identity.invert` is inlined them most of the loop body becomes invariant and gets hoisted out leaving just allocation of a new matrix behind. 

This is the difference between JIT and AOT - JIT inlines, AOT does not. You can put `@pragma('vm:never-inline')` on the `invert` to compare actual performance.

Note: SIMD API that Dart has was designed with Intel in mind and some of the operations (e.g. shuffle) don't have a single instruction lowering on ARM, instead they are lowered into multiple instructions and thus might be rather slow. This more or less means that you have to stay within portable subset of the SIMD if you want to write efficient code.

```
        ;; v564 <- SimdOp:176(Float32x4Shuffle, mask = 177, v562) T{_Float32x4}
0x7f8597119b90    6e0424a3               vinss v3[0], v5[1]
0x7f8597119b94    6e0c04a3               vinss v3[1], v5[0]
0x7f8597119b98    6e1464a3               vinss v3[2], v5[3]
0x7f8597119b9c    6e1c44a3               vinss v3[3], v5[2]
```

There is also something very strange going on with spilling, we spill the same SSA value twice into the same location

```
        ;; v604 <- SimdOp:192(Float32x4Mul, v598, v360) T{_Float32x4}
0x7f8597119bcc    6e2adc64               vmuls v4, v3, v10
        ;; ParallelMove QS-8 <- v4
0x7f8597119bd0    3c9c03a4               fstrq v4, [fp, #-64]
        ;; v610 <- SimdOp:194(Float32x4Mul, v598, v536) T{_Float32x4}
0x7f8597119bd4    6e22dc65               vmuls v5, v3, v2
        ;; ParallelMove QS-6 <- v5
0x7f8597119bd8    3c9d03a5               fstrq v5, [fp, #-48]
        ;; v616 <- SimdOp:196(Float32x4Mul, v598, v490) T{_Float32x4}
0x7f8597119bdc    6e20dc62               vmuls v2, v3, v0
        ;; ParallelMove QS-4 <- v2
0x7f8597119be0    3c9e03a2               fstrq v2, [fp, #-32]
        ;; v622 <- SimdOp:198(Float32x4Mul, v598, v548) T{_Float32x4}
0x7f8597119be4    6e21dc60               vmuls v0, v3, v1
        ;; ParallelMove QS-2 <- v0
0x7f8597119be8    3c9f03a0               fstrq v0, [fp, #-16]
        ;; ParallelMove QS-8 <- v4, QS-6 <- v5, QS-4 <- v2, QS-2 <- v0
0x7f8597119bec    3c9c03a4               fstrq v4, [fp, #-64]
0x7f8597119bf0    3c9d03a5               fstrq v5, [fp, #-48]
0x7f8597119bf4    3c9e03a2               fstrq v2, [fp, #-32]
0x7f8597119bf8    3c9f03a0               fstrq v0, [fp, #-16]
        ;; v198 <- AllocateObject(SimdMatrix4 <not-aliased>) T{SimdMatrix4}
0x7f8597119bfc    94000000               bl 0x7f8597119bfc
```

There is also a question if we should be spilling FPU registers at all across `AllocateObject` - we should probably move these FPU spills into the slow path of the allocation. WDYT @mkustermann? 

I have been looking around and found [this library](https://github.com/projectNe10/Ne10/blob/master/modules/math/NE10_invmat.neon.s) written by ARM engineers - it seems whatever we do we should probably base it on that. 

 /cc @askeksa-google who has experience with ARM / SIMD and might have some input on this.

**Comment by mkustermann** [comment]
> There is also a question if we should be spilling FPU registers at all across AllocateObject - we 
> should probably move these FPU spills into the slow path of the allocation. WDYT 
> @mkustermann?

Agree. Additionally we could look into avoiding `null` initializing the data in the generic allocation stub if the field are unboxed (in AOT).

**Comment by jonahwilliams** [comment]
I did not know about the `never-inline` pragma, thanks @mraleph . If I update the code so that both the SIMD invert and the vector_math invert are annotated with it I get very different numbers.

JIT:
SIMD: ~0.11 us / invert
vector_math:  ~0.08 us /invert

AOT:
SIMD:  ~ ~0.11 us /invert
vector_math: ~ 0.08 us /invert

So it seems like all of the performance difference between  AOT/JIT was due to more aggressive in-lining. Woops! 

>Note: SIMD API that Dart has was designed with Intel in mind and some of the operations (e.g. shuffle) don't have a single instruction lowering on ARM

That is unfortunate, but good to know. In that case, I'd have to read up a bit on an efficient implementation for ARM.

**Comment by jonahwilliams** [comment]
Also fixing the in-lining and running via dart2native on my desktop,

AOT:
SIMD:  ~0.01 us / invert
vector_math:  ~0.05 us/ invert

I guess I would need arch-specific invert implementations

**Comment by rakudrama** [comment]
You probably get much much worse performance on the web for SIMD since it is not supported in JavaScript and faked with lots of temporary FloatNNList(4) objects.
If you use SIMD in code common to native and web you will need to refactor it to allow conditional imports.

**Comment by jonahwilliams** [comment]
@rakudrama Yes, we'd definitely want to make sure that we don't switch to something that is slower on the web. To that end, we could conditional imports or some sort of tree-shakeable branching

I pushed a new update with a different implementation of matrix inversion that seems to perform a lot better, running about 2/3 the speed of the vector math implementation on a Pixel 4. https://github.com/jonahwilliams/flutter/commit/cfa93c8ae6e0ef5f82d0071d74d2b65eae45825f , though I have only verified on the identity matrix!

I don't have a good way to transpose the matrix at the end, putting it together without any SIMD operations doesn't seem as slow as I thought.
